# Supervised Learning

## Example

- Does a specific image contain a person's face?
- Training data: vectors of pixel values
- Labels: 1 or 0

## Classification Problems

### Binary classification

> Will a person purchase the insurance package given some quote?
 
**AUC**: Metric for binary classification models.

- Area under the ROC curve (AUC)
- Larger area under the ROC curve = better model

### Multi-class classification

> What is the species of a given bird?

Accuracy score and confusion matrix are common metrics.

- Confusion matrix


|      Actual      | Predicted: Spam Email | Predicted: Real Email |
|:----------------:|:---------------------:|:---------------------:|
|    Spam Email    |     True Positive     |    False Positive     |
|    Real Email    |    False Positive     |     True Positive     |


- Accuracy

$$\frac{tp + tn}{tp + tn + fp + fn}$$

Other supervised learning considerations
- Features (Also known as Attributes or Predictors) can be either numeric or categorical
- Numeric features should be scaled (Z-scored)
- Categorical features should be encoded (One-hot encoding)

**Ranking problems**
- Predicting an ordering on a set of choices

**Recommendation problems**
- Recommending a product or service to a user
- Based on consumption history and profile

## What is XGBoost

- Optimized gradient-boosting machine learning library
- Originally written in C++
- Has APIs in several languages:
   - **Python**
   - R
   - Julia
   - Scala
   - Java

### What makes XGBoost so popular?

- Speed and performance
- Core algorithm is parallelizable
- Consistently outperforms other single-algorithm methods
- State-of-the-art performance in many ML tasks

### Using XGBoost: a quick example.

In [ ]:
import numpy as np
import pandas as pd
import xgboost as xgb
from sklearn.model_selection import train_test_split

# Load the data
class_data = pd.read_csv('class_data.csv')
# Split the entire dataset into features (X) and target (y)
X, y = class_data.iloc[:, :-1], class_data.iloc[:, -1]
# Split the dataset into a training and a test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)
# Instantiate the XGBClassifier
xg_cl = xgb.XGBClassifier(objective='binary:logistic', n_estimators=10, seed=123)
# Fit the classifier to the training set
xg_cl.fit(X_train, y_train)
# Predict the labels of the test set
preds = xg_cl.predict(X_test)
# Compute the accuracy: accuracy
accuracy = float(np.sum(preds == y_test)) / y_test.shape[0]
print("accuracy: %f" % (accuracy))

### What is a decision tree?

Because XGBoost is an ensemble method, it relies on the use of many decision trees to make predictions. So, what is a decision tree?

Each node contains a question, with only two possible choices. At the very bottom of the tree, there is a single possible decision. Every possible path will lead to a decision.

#### Decision trees as base learners

Base learner - individual learning algorithm in an ensemble algorithm
Composed of a series of binary questions
Predictions happen at the "leaves" of the tree

### Decision trees and CART

Constructed iteratively (one decision at a time), until a stopping criterion is met

#### Individual decision trees

Are low bias, high variance in general. This means that they are very good at learning relationships but they tend to overfit the training data. And often they do not generalize well to new data.

#### CART: Classification and Regression Trees

Each leaf **always** contains a real-valued score regardless of whether the tree is used for classification or regression
The real-valued scores can later be tresholded to convert into categories for classification problems if necessary


In [ ]:
# Import the necessary modules
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

# Load the data
class_data = pd.read_csv('class_data.csv')
# Split the entire dataset into features (X) and target (y)
X, y = class_data.iloc[:, :-1], class_data.iloc[:, -1]
# Create the training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=123)

# Instantiate the classifier: dt_clf_4
dt_clf_4 = DecisionTreeClassifier(max_depth=4)

# Fit the classifier to the training set
dt_clf_4.fit(X_train, y_train)

# Predict the labels of the test set: y_pred_4
y_pred_4 = dt_clf_4.predict(X_test)

# Compute the accuracy of the predictions: accuracy
accuracy = float(np.sum(y_pred_4 == y_test)) / y_test.shape[0]
print("accuracy:", accuracy)


### What is Boosting?

- Not a specific machine learning algorithm
- Concept that can be applied to a set of machine learning models: "Meta-algorithm"
- Ensemble meta-algorithm used to convert many weak learners into a strong learner

#### Weak learners and strong learners

- **Weak learner**: ML algorithm that is slightly better than chance
    - Example: Decision tree whose predictions are slightly better than 50%
    - Boosting converts a collection of weak learners into a strong learner
- **Strong learner**: Any algorithm that can be tuned to achieve good performance

#### How boosting is accomplished

- Iteratively learning a set of weak models on subsets of the data
- Weighing each weak prediction according to each weak learner's performance
- Combine the weighted predictions to obtain a single weighted prediction... that is much stronger than the individual ones

#### Model evaluation through cross-validation

- Cross-validation: Robust method for estimating the performance of a model on unseen data
- Generates many none-overlapping train/test splits on training data
- Reports the average test set performance across all data splits


In [ ]:
import xgboost as xgb
import pandas as pd

churn_data = pd.read_csv('churn_data.csv')

churn_dmatrix = xgb.DMatrix(data=churn_data.iloc[:, :-1], label=churn_data.month_5_still_here)
params = {"objective": "binary:logistic", "max_depth": 4}
# nfold: number of cross-validation folds
# num_boost_round: number of trees we build
# metrics: metric to be computed (error for accuracy)
cv_results = xgb.cv(dtrain=churn_dmatrix, params=params, nfold=4, num_boost_round=10, metrics="error", as_pandas=True)
print("Accuracy: %f" % ((1 - cv_results["test-error-mean"]).iloc[-1]))

### When to use XGBoost

- Large number of training samples
    - Greater than 1000 training samples and less than 100 features
    - The number of features < number of training samples 
- You have a mixture of categorical and numeric features
  - Or just numeric features

### When to NOT use XGBoost
- Image recognition
- Computer vision
- Natural language processing and understanding problems
- When the number of training samples is significantly smaller than the number of features

## Demo

We will be working with the Diamonds dataset throughout the tutorial. It is built into the `Seaborn` library, or alternatively, you can also download it from [Kaggle](https://www.kaggle.com/datasets/shivam2503/diamonds?resource=download). It has a nice combination of numeric and categorical features and over 50k observations that we can comfortably showcase all the advantages of XGBoost.



In [1]:
import seaborn as sns
import warnings

warnings.filterwarnings("ignore")

diamonds = sns.load_dataset("diamonds")

diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [2]:
diamonds.shape

(53940, 10)

In a typical real-world project, you would want to spend a lot more time exploring the dataset and visualizing its features. But since this data comes built-in to Seaborn, it is relatively clean.

So, we will just look at the 5-number summary of the numeric and categorical features and get going. You can spend a few moments to familiarize yourself with the dataset.

In [3]:
diamonds.describe()

carat         depth         table         price             x  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940     61.749405     57.457184   3932.799722      5.731157   
std        0.474011      1.432621      2.234491   3989.439738      1.121761   
min        0.200000     43.000000     43.000000    326.000000      0.000000   
25%        0.400000     61.000000     56.000000    950.000000      4.710000   
50%        0.700000     61.800000     57.000000   2401.000000      5.700000   
75%        1.040000     62.500000     59.000000   5324.250000      6.540000   
max        5.010000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53940.000000  53940.000000  
mean       5.734526      3.538734  
std        1.142135      0.705699  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.040000  
max       58.900000     31.800000

In [4]:
import numpy as np

diamonds.describe(exclude=np.number)

cut  color clarity
count   53940  53940   53940
unique      5      7       8
top     Ideal      G     SI1
freq    21551  11292   13065

### Preprocessing

After you are done with exploration, the first step in any project is framing the machine learning problem and extracting the feature and target arrays based on the dataset.

In this tutorial, we will first try to predict diamond prices using their physical measurements, so our target will be the price column.

So, we are isolating the features into X and the target into y.


In [5]:
# Extract feature and target arrays
X, y = diamonds.drop('price', axis=1), diamonds[['price']]

# Extract text features
cats = X.select_dtypes(exclude=np.number).columns.tolist()

# Convert to Pandas category
for col in cats:
    X[col] = X[col].astype('category')

X.dtypes

carat       float64
cut        category
color      category
clarity    category
depth       float64
table       float64
x           float64
y           float64
z           float64
dtype: object

In [6]:
import xgboost as xgb

from sklearn.model_selection import train_test_split

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.25, random_state=123)

# Create regression matrices
dtrain_reg = xgb.DMatrix(X_train, y_train, enable_categorical=True)
dtest_reg = xgb.DMatrix(X_test, y_test, enable_categorical=True)

After building the DMatrices, you should choose a value for the `objective` parameter. It tells XGBoost the machine learning problem you are trying to solve and what metrics or loss functions to use to solve that problem.

For example, to predict diamond prices, which is a regression problem, you can use the common `reg:squarederror` objective. Usually, the name of the objective also contains the name of the loss function for the problem. For regression, it is common to use Root Mean Squared Error, which minimizes the square root of the squared sum of the differences between actual and predicted values. Here is how the metric would look like when implemented in NumPy:


```python
import numpy as np

mse = np.mean((actual - predicted) ** 2)
rmse = np.sqrt(mse)
```

We’ll learn classification objectives later in the tutorial.

A note on the difference between a loss function and a performance metric: A loss function is used by machine learning models to minimize the differences between the actual (ground truth) values and model predictions. On the other hand, a metric (or metrics) is chosen by the machine learning engineer to measure the similarity between ground truth and model predictions.

In short, a loss function should be minimized while a metric should be maximized. A loss function is used during training to guide the model on where to improve. A metric is used during evaluation to measure overall performance.

### Training the model

Inside this initial `params`, we are also setting `tree_method` to `gpu_hist`, which enables GPU acceleration. If you don't have a GPU, you can omit the parameter or set it to `hist`.

Now, we set another parameter called `num_boost_round`, which stands for number of boosting rounds. Internally, XGBoost minimizes the loss function RMSE in small incremental rounds (more on this later). This parameter specifies the amount of those rounds.

The ideal number of rounds is found through hyperparameter tuning. For now, we will just set it to 100:

In [7]:
# Define hyperparameters
params = {"objective": "reg:squarederror", "tree_method": "gpu_hist"}
n = 100
model = xgb.train(
    params=params,
    dtrain=dtrain_reg,
    num_boost_round=n,
)

In [8]:
from sklearn.metrics import mean_squared_error

preds = model.predict(dtest_reg)
rmse = mean_squared_error(y_test, preds, squared=False)

print(f"RMSE of the base model: {rmse:.3f}")

RMSE of the base model: 542.265


### Evaluating the model

Training a machine learning model is like launching a rocket into space. You can control everything about the model up to the launch, but once it does, all you can do is stand by and wait for it to finish.

But the problem with our current training process is that we can’t even watch where the model is going. To solve this, we will use evaluation arrays that allow us to see model performance as it gets improved incrementally across boosting rounds.

First, let’s set up the parameters again:

In [9]:
params = {"objective": "reg:squarederror", "tree_method": "gpu_hist"}
n = 100

Next, we create a list of two tuples that each contain two elements. The first element is the array for the model to evaluate, and the second is the array’s name.

In [10]:
evals = [(dtrain_reg, "train"), (dtest_reg, "validation")]

When we pass this array to the evals parameter of xgb.train, we will see the model performance after each boosting round:

In [11]:
model = xgb.train(
    params=params,
    dtrain=dtrain_reg,
    num_boost_round=n,
    evals=evals,
)

[0]	train-rmse:2862.31146	validation-rmse:2840.68834
[1]	train-rmse:2082.81862	validation-rmse:2065.21072
[2]	train-rmse:1549.44507	validation-rmse:1536.71389
[3]	train-rmse:1187.07459	validation-rmse:1176.80318
[4]	train-rmse:945.91339	validation-rmse:939.61269
[5]	train-rmse:790.74365	validation-rmse:791.39691
[6]	train-rmse:691.02783	validation-rmse:696.96497
[7]	train-rmse:630.99900	validation-rmse:644.34523
[8]	train-rmse:590.14399	validation-rmse:607.29011
[9]	train-rmse:564.79697	validation-rmse:585.92024
[10]	train-rmse:548.30676	validation-rmse:572.75567
[11]	train-rmse:536.77530	validation-rmse:564.87578
[12]	train-rmse:529.14333	validation-rmse:559.05668
[13]	train-rmse:522.04003	validation-rmse:554.92832
[14]	train-rmse:516.57524	validation-rmse:551.69600
[15]	train-rmse:510.55122	validation-rmse:548.89999
[16]	train-rmse:507.40026	validation-rmse:548.23401
[17]	train-rmse:501.95883	validation-rmse:547.28474
[18]	train-rmse:497.47550	validation-rmse:545.46888
[19]	train-rms

In [12]:
n = 5000

evals = [(dtrain_reg, "train"), (dtest_reg, "validation")]

model = xgb.train(
    params=params,
    dtrain=dtrain_reg,
    num_boost_round=n,
    evals=evals,
    verbose_eval=500
)

[0]	train-rmse:2862.31146	validation-rmse:2840.68834
[500]	train-rmse:195.16395	validation-rmse:562.24713
[1000]	train-rmse:119.62424	validation-rmse:570.01617
[1500]	train-rmse:82.38357	validation-rmse:572.87333
[2000]	train-rmse:60.83733	validation-rmse:575.02089
[2500]	train-rmse:46.95207	validation-rmse:576.43277
[3000]	train-rmse:36.08419	validation-rmse:577.29667
[3500]	train-rmse:29.11954	validation-rmse:577.78109
[4000]	train-rmse:23.75240	validation-rmse:578.22335
[4500]	train-rmse:20.01431	validation-rmse:578.47206
[4999]	train-rmse:17.15337	validation-rmse:578.63708


In [13]:
n = 10000

evals = [(dtrain_reg, "train"), (dtest_reg, "validation")]

model = xgb.train(
    params=params,
    dtrain=dtrain_reg,
    num_boost_round=n,
    evals=evals,
    verbose_eval=50,
    # Activate early stopping
    early_stopping_rounds=150
)

[0]	train-rmse:2862.31146	validation-rmse:2840.68834
[50]	train-rmse:430.19810	validation-rmse:542.52170
[100]	train-rmse:372.89833	validation-rmse:541.72131
[150]	train-rmse:327.91646	validation-rmse:549.48379
[181]	train-rmse:309.51202	validation-rmse:550.20288


### XGBoost Cross-validation


In [14]:
params = {"objective": "reg:squarederror", "tree_method": "gpu_hist"}
n = 1000

results = xgb.cv(
    params, dtrain_reg,
    num_boost_round=n,
    nfold=5,
    early_stopping_rounds=20
)

results.head()

train-rmse-mean  train-rmse-std  test-rmse-mean  test-rmse-std
0      2863.735582        5.736820     2866.800179      28.367405
1      2083.181962        5.689362     2089.536563      19.794082
2      1547.302833        4.178947     1557.401456      17.321109
3      1183.037644        3.214807     1198.799702      16.904746
4       941.048091        3.030368      961.480079      15.845508

In [15]:
best_rmse = results['test-rmse-mean'].min()
print(f"Best RMSE: {best_rmse:.2f}")

Best RMSE: 548.75


Note that this method of cross-validation is used to see the true performance of the model. Once satisfied with its score, you must retrain it on the full data before deployment.

In [17]:
from sklearn.model_selection import GridSearchCV
import xgboost as xgb

param_grid = {
    "learning_rate": [0.05, 0.10, 0.15, 0.20, 0.25, 0.30],
    "max_depth": [3, 4, 5, 6, 8, 10, 12, 15],
    "min_child_weight": [1, 3, 5, 7],
    "gamma": [0.0, 0.1, 0.2, 0.3, 0.4],
    "colsample_bytree": [0.3, 0.4, 0.5, 0.7]
}

model = xgb.XGBRegressor()
search = GridSearchCV(model, param_grid, cv=5)

search.fit(X_train.drop(columns=cats), y_train)

print(search.best_params_)

{'colsample_bytree': 0.7, 'gamma': 0.0, 'learning_rate': 0.05, 'max_depth': 8, 'min_child_weight': 7}


### Hyperparameter Tuning

Values of hyperparameters that maximize the model's performance.

```json
{
    "colsample_bytree": 0.7,
    "gamma": 0.0,
    "learning_rate": 0.05,
    "max_depth": 8,
    "min_child_weight": 7
}
```

- `colsample_bytree`: `0.7`: This means that each tree will randomly select 70% of the features before it starts to grow. This level of randomness helps prevent overfitting.
- `gamma`: `0.0`: This is the specified minimum loss reduction required to make a further partition on a leaf node of the tree. A value of 0.0 indicates that there is no minimum loss reduction required, which may lead to more complex trees.
- `learning_rate`: `0.05`: This is your learning rate (or eta). It signifies a relatively slower learning rate. Slower learning rates often lead to better performance but are slower to train.
- `max_depth`: `8`: This is the maximum depth per tree. Deeper trees can model more complex relationships by adding more nodes, but they also risk overfitting.
- `min_child_weight`: `7`: This is the minimum sum of instance weight (hessian) needed in a child. This parameter is used to control overfitting. The larger the value, the more conservative the algorithm will be. A value of 7 means that splitting a node will only get considered if it contains at least seven instances weighted by their Hessian (second order gradient).

In [32]:
optimal_params = {'colsample_bytree': 0.7,
                  'gamma': 0.0,
                  'learning_rate': 0.05,
                  'max_depth': 8,
                  'min_child_weight': 7,
                  'objective': 'reg:squarederror'}
n = 10_000

model = xgb.train(
    params=optimal_params,
    dtrain=dtrain_reg,
    num_boost_round=n,
    evals=evals,
    verbose_eval=50,
    # Activate early stopping
    early_stopping_rounds=150)


[0]	train-rmse:3801.30313	validation-rmse:3784.05907
[50]	train-rmse:620.07539	validation-rmse:651.29847
[100]	train-rmse:451.56478	validation-rmse:526.75196
[150]	train-rmse:421.70235	validation-rmse:520.71008
[200]	train-rmse:400.34397	validation-rmse:519.80790
[250]	train-rmse:387.77836	validation-rmse:520.10523
[300]	train-rmse:377.73299	validation-rmse:520.13723
[350]	train-rmse:369.35932	validation-rmse:520.40862
[368]	train-rmse:366.30248	validation-rmse:520.98613


And we can see base score this again:

In [33]:
from sklearn.metrics import mean_squared_error

# Generate predictions
preds = model.predict(dtest_reg)
rmse = mean_squared_error(y_test, preds, squared=False)

print(f"RMSE of the base model: {rmse:.3f}")

RMSE of the base model: 521.023


The Root Mean Squared Error value of `521.023` means that, on average, the model's predicted values are approximately `521` units away from the actual values.

In [34]:
results = xgb.cv(
    params=optimal_params,
    dtrain=dtrain_reg,
    num_boost_round=n,
    nfold=5,
    early_stopping_rounds=20
)

best_rmse = results['test-rmse-mean'].min()
print(f"Best RMSE: {best_rmse:.3f}")

Best RMSE: 543.421


The 'best RMSE' of `543.421` was the best score observed during the CV process, meaning it was the smallest average RMSE across testing on all different folds. 

This is often a more reliable indicator of how your model will perform on unseen data, because it averages over several different splits of the data rather than relying on a single split.

In [36]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test, preds)
print(f'R-Squared: {r2:.5f}')

R-Squared: 0.98283
